### PR TITLE
[`arithmetic_side_effects`] Fix #11393

### LINKED ISSUE #11393: [arithmetic_side_effects] Division by potentially-zero can panic even with `Wrapping`
### Summary

Division (`/` and `%`, i.e. `div()` and `rem()`) of any `core::num::Wrapping` by `Wrapping(0)` (of the same type) panics, but the `arithmetic_side_effects` lint currently doesn't flag them (as opposed to the same operations with primitive integers).

(Note: kind of the opposite of #11392)

### Lint Name

arithmetic_side_effects

### Reproducer

I tried this code:

```rust
#![warn(clippy::arithmetic_side_effects)]

use std::num::Wrapping;

fn example_div(x: Wrapping<i32>, maybe_zero: Wrapping<i32>) -> Wrapping<i32> {
    x / maybe_zero
}

fn example_rem(x: Wrapping<i32>, maybe_zero: Wrapping<i32>) -> Wrapping<i32> {
    x % maybe_zero
}

fn main() {
    use std::panic;

    let [x, maybe_zero] = [1, 0].map(Wrapping);

    println!("{x} / {maybe_zero} is {:?}", panic::catch_unwind(|| example_div(x, maybe_zero)));
    println!("{x} % {maybe_zero} is {:?}", panic::catch_unwind(|| example_rem(x, maybe_zero)));
}
```

(note: compiles but panics with "`attempt to divide by zero`" and "`attempt to calculate the remainder with a divisor of zero`")

I expected to see this happen:

```
warning: arithmetic operation that can potentially result in unexpected side-effects
 --> src/main.rs:6:5
  |
6 |     x / maybe_zero
  |     ^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::arithmetic_side_effects)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: arithmetic operation that can potentially result in unexpected side-effects
  --> src/main.rs:10:5
   |
10 |     x % maybe_zero
   |     ^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

Instead, this happened:

No warning

### Version

```text
Clippy 0.1.73 (2023-08-23 249595b) on the Rust Playground,
which I guess corresponds to rustc 1.74.0-nightly (2023-08-23 249595b7523fc07a99c1)
```

### PR DESCRIPTION
**Pull request by c410-f3r**
#11395 tried to fix #11393 but it actually didn't due to an oversight on my side. Now everything should be OK.

```
changelog: [`arithmetic_side_effects`]: Detect division by zero for `Wrapping` and `Saturating`
```

### CODE REVIEW

File: clippy_lints/src/operators/arithmetic_side_effects.rs [hunk 1]
```diff
@@ -91,36 +91,24 @@ impl ArithmeticSideEffects {
...
+
+        // If the RHS is NonZeroU*, then division or module by zero will never occur
+        if is_non_zero_u(type_diagnostic_name(cx, rhs_ty)) && is_div_or_rem {
             return true;
         }
-        // For `Saturation` or `Wrapping` (LHS), everything is allowed
-        if match_type(cx, lhs_ty, SATURATING) || match_type(cx, lhs_ty, WRAPPING) {
-            return true;
+        // `Saturation` and `Wrapping` can overflow if the RHS is zero in a division or module
+        if is_sat_or_wrap(lhs_ty) || is_sat_or_wrap(rhs_ty) {
```
Comment (guilliamxavier) [inline clippy_lints/src/operators/arithmetic_side_effects.rs:108]: Actually it could/should only check the LHS (like before the previous PR)?

Comment (c410-f3r) [reply clippy_lints/src/operators/arithmetic_side_effects.rs:108]: Done

**Comment by c410-f3r** [comment]
r? clippy

**Comment by flip1995** [comment]
AFAICT this PR is superseded by the PR you just opened in the Rust repo?

**Comment by c410-f3r** [comment]
They are almost exactly the same to reduce any potential rebasing conflicts. So it is more like the PR in the Rust repository is the "canonical" version because of the symbols.

**Comment by flip1995** [comment]
Ok, thanks for the clarification. Then I'd rather merge the one in the Rust repo. I don't want to do a whole sync dance (over 3 weeks) around this fix.

Closing as superseded by https://github.com/rust-lang/rust/pull/115415

**Pull request closed without merging**